test/hooks.on-listen.test.js

```diff
@@ -0,0 +1,614 @@
+  })
+})
+t.test('nonlocalhost onListen encapsulation should be called in order', t => {
+  t.plan(6)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, 'called in root')
+    t.equal(this.pluginName, fastify.pluginName, 'the this binding is the right instance')
+    done()
+  })
+
+  fastify.register(async (childOne, o) => {
+    childOne.addHook('onListen', function (done) {
+      t.equal(order++, 1, 'called in childOne')
+      t.equal(this.pluginName, childOne.pluginName, 'the this binding is the right instance')
+      done()
+    })
+    childOne.register(async (childTwo, o) => {
+      childTwo.addHook('onListen', async function () {
+        t.equal(order++, 2, 'called in childTwo')
+        t.equal(this.pluginName, childTwo.pluginName, 'the this binding is the right instance')
+      })
+    })
+  })
+  fastify.listen({
+    host: '::1',
+    port: 0
+  })
+})
+t.test('onListen localhost should work in order with callback', t => {
+  t.plan(4)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, '1st called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 1, '2nd called in root')
+    done()
+  })
+
+  fastify.listen({ port: 0 }, (err) => {
+    t.equal(fastify.server.address().address, localhost)
+    t.error(err)
+  })
+})
+t.test('onListen localhost should work in order with callback in async', t => {
+  t.plan(4)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 0, '1st called in root')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 1, '2nd called in root')
+  })
+
+  fastify.listen({ host: 'localhost', port: 0 }, (err) => {
+    t.equal(fastify.server.address().address, localhost)
+    t.error(err)
+  })
+})
+t.test('onListen localhost with callback should manage error in sync', t => {
+  t.plan(5)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.pass('1st called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, 'error sync called in root')
+    done(new Error('FAIL ON LISTEN'))
+  })
+
+  fastify.listen({ port: 0 }, (err) => {
+    if (err) {
+      t.ok(err, 'FAIL ON LISTEN')
+    } else {
+      t.equal(fastify.server.address().address, localhost)
+      t.error(err)
+    }
+  })
+})
+t.test('onListen localhost with callback should manage error in async', t => {
+  t.plan(5)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', async function () {
+    t.pass('1st called in root')
```

Comment:
dangerous. explicitly marking the test as passing

Reply:
We were trying to follow `test\hooks.on-ready.test.js` as a reference. We might have done so wrong.

Reply:
maybe on-ready.test.js is also problematic. will have a look

Reply:
nevermind regarding t.pass, it only increments the count.